### ISSUE TITLE
ui: request in statements page causes gRPC error due to "size too large"

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#54439]
Reported by @onno-vos-dev today:

![Screenshot at 2020-09-16 10-11-51](https://user-images.githubusercontent.com/11063313/93310638-556db700-f805-11ea-8d8d-f82331a5e713.png)

A statement page is not displayed and instead gRPC complains about a 5GB payload.

We need to figure two things out:

- what caused the "5GB" mentioned in the erorr message: is that an actual size, or is this some other quantity that's mis-represented in the error?
- what would block the statements page from displaying?
- what alternatives (workarounds) are there?

The user also mentions #54022 as an adjacent issue that could help power the statements page when there are many statements.

**Comment by knz** [comment]
@onno-vos-dev if you could provide the following that would be swell:

- cockroachdb version (and platform)
- whether you can run `show cluster queries` at all in a SQL shell
  - if you can, `select count(*) from [show cluster queries]` so we get an estimate of how much data you're working with
- what is the value of the cluster settings:
  - `diagnostics.forced_sql_stat_reset.interval`
  - `diagnostics.sql_stat_reset.interval`
  - `diagnostics.reporting.enabled`
  - `diagnostics.reporting.interval`
- whether you can run `table crdb_internal.node_statement_statistics` in the SQL shell at all
  - if you can, `select count(*) from crdb_internal.node_statement_statistics`

Thank you

**Comment by onno-vos-dev** [comment]
@knz Here you go

### CockroachDB version:
```
coc[onno.vos@crdb1 ~]$ cockroach version
Build Tag:        v20.2.0-alpha.1-3163-g8b9e8dc32e
Build Time:       2020/09/07 11:53:15
Distribution:     CCL
Platform:         linux amd64 (x86_64-redhat-linux)
Go Version:       go1.13.14
C Compiler:       gcc 4.8.5 20150623 (Red Hat 4.8.5-39)
Build Commit ID:  8b9e8dc32e73cfdfc1999da35d61e5cc9a2b35ec
Build Type:       development
```
(Basically we're running `commit 8b9e8dc32e73cfdfc1999da35d61e5cc9a2b35e`) on CentOS Linux release 7.8.2003 (Core)

### show cluster queries
```
root@10.***.***.***:26257/defaultdb> show cluster queries;
              query_id             | node_id |            session_id            | user_name |              start               |        query         |    client_address    | application_name | distributed |   phase
-----------------------------------+---------+----------------------------------+-----------+----------------------------------+----------------------+----------------------+------------------+-------------+------------
  163537bf2acb50ca0000000000000001 |       1 | 163537bebc0c18210000000000000001 | root      | 2020-09-16 08:55:36.796264+00:00 | SHOW CLUSTER QUERIES | 10.***.***.***:**** | $ cockroach sql  |    false    | executing
(1 row)

Server Execution Time: 8.291ms
Network Latency: 271.45µs
```

```
root@10.***.***.***:26257/defaultdb> select count(*) from [show cluster queries];
  count
---------
      1
(1 row)

Server Execution Time: 8.992ms
Network Latency: 196.203µs
```

### cluster settings
```
diagnostics.forced_sql_stat_reset.interval -> 23h0m0s
diagnostics.sql_stat_reset.interval -> 23h0m0s 
diagnostics.reporting.enabled -> true 
diagnostics.reporting.interval -> 23h0m0s 
```

### node_statement_statistics
I can run `table crdb_internal.node_statement_statistics` from a SQL-shell and get the following execution times:
```
Server Execution Time: 10.597ms
Network Latency: 1.437362ms
```

```
root@10.***.***.***:26257/defaultdb> select count(*) from crdb_internal.node_statement_statistics;
  count
---------
     53
(1 row)

Server Execution Time: 10.82ms
Network Latency: 244.53µs
```

Figured the following output might be of interest as well:
```
root@10.***.***.***:26257/defaultdb> select count(*) from crdb_internal.node_transaction_statistics;
  count
---------
  58608
(1 row)

Server Execution Time: 1.54673s
Network Latency: 239.197µs
```

**Comment by knz** [comment]
Yes these 50000 rows worth of transaction statistics are no good.

It looks like your app is auto-generating SQL, whereby CockroachDB has the impression that every query is different (so there is little to no aggregation of stats between queries -- we try to aggregate stats when queries "look the same" otherwise).

If your app indeed cuases so many variations of SQL, it's going to be on you to reduce the reset interval down from 23 hours to something that keeps the number of entries there under, say, 5000 rows.

**Comment by knz** [comment]
@piyush-singh customer reports that the application/DB legitimately sends many diverse queries and so 50k+ entries in the statements page is going to be common. 

My opinion is that we should not depend on manual configuration to ensure the system does not "blow up" in this use case.

The proper technical design would ensure that only the most relevant rows are transmitted from the crdb back-end (SQL) to the admin UI. This connects to #54022 somewhat but I feel the statements page should receive a dedicated solution.

Maybe @jordanlewis @dhartunian would have suggestions about how to go about achieving this.

**Comment by jordanlewis** [comment]
Looping in @solongordon. We might need to do something here before release to protect against memory issues.

**Comment by arulajmani** [comment]
There's https://github.com/cockroachdb/cockroach/issues/54452 that I found today, which would account for the high number of transactions stored.

**Comment by arulajmani** [comment]
This issue in particular seems to have been caused by #54452, but in theory we are still susceptible to this problem if the `appStats` datastructure grows too large. I have https://github.com/cockroachdb/cockroach/pull/54521 out for memory accounting, but it adds more contention on `appStats`, which will presumably exacerbate #54515 (which is under investigation currently).

**Comment by awoods187** [comment]
Is there an update on this issue?

**Comment by solongordon** [comment]
The bug which likely caused this particular issue (#54452) was resolved for v20.2.0. However it is still theoretically possible for statistics to grow too large and caused this problem, especially for users who set a long reporting period. We believe Arul's other PR (#54521) will address this problem. It was initially set aside while we addressed some performance concerns with transaction statistics gathering so did not make it into v20.2.0. But now that those performance concerns have been addressed, we'll review that PR and see if it is suitable for including in the next patch release.
